src/Controller/PostfinanceResponseController.php

```diff
@@ -25,27 +26,42 @@ class PostfinanceResponseController {
    * (pending, status 51 or 91).
    *
    * @param \Symfony\Component\HttpFoundation\Request $request
-   *   Request
+   *   Request.
    * @param \Drupal\payment\Entity\PaymentInterface $payment
    *   The Payment Entity type.
+   *
+   * @return \Drupal\Core\Response
+   *   The Response to the accepting request.
    */
   public function processAcceptResponse(Request $request, PaymentInterface $payment) {
     // The definition of the plugin implementation.
     $plugin_definition = $payment->getPaymentMethod()->getPluginDefinition();
+    try {
+      $request_data = $request->query->all();
+      $sha_sent = $request_data['SHASIGN'];
+      unset($request_data['SHASIGN']);
+
+      // Generate SHASign from request data.
+      $sha_sign = PostfinanceHelper::generateShaSign($request_data, $plugin_definition['sha_out_key']);
 
-    // Generate local SHASign.
-    $request_values = $request->query->all();
+      // Check if the sent signature is valid.
+      if ($sha_sign != $sha_sent) {
+        throw new \Exception('Invalid postfinance key.');
+      }
 
-    // Generate SHA-OUT signature
-    $sha_sign = PostfinanceHelper::generateShaSign($request_values, $plugin_definition['sha_out_key']);
+      if ($request_data['STATUS'] == 'error') {
+        throw new \Exception('There was an error processing the request:' . $request_data['NCERROR']);
+      }
 
-    if ($sha_sign == $request->get('SHASIGN')) {
-      drupal_set_message(t('Payment succesfull.'), 'error');
-      $this->savePayment($payment, 'payment_success');
-    } else {
-      $this->savePayment($payment, 'payment_failed');
-      \Drupal::logger(t('Payment verification failed: @error'),array('@error' => 'SHASign did not equal'))->warning('PostfinanceResponseController.php');
+      if ($request_data['STATUS'] == 5 || $request_data['STATUS'] == 9) {
+        drupal_set_message(t('Payment successful.'));
+        return $this->savePayment($payment, 'payment_success');
+      }
+    }
+    catch (\Exception $e) {
+      \Drupal::logger(t('Payment verification failed: @error'), array('@error' => 'SHASign did not equal'))->warning('PostfinanceResponseController.php');
```

Comment:
Something like this is what we actually want to do for the error cases above, but we will never get here, because that's not how payment validation fails for postfinance.